### PR TITLE
Support cgroup2

### PR DESCRIPTION
**Pull request by emadolsky**
CGroup-V2 has a different architecture which is unified. Since it is not yet supported in go, it makes sense to currently have the
support. This commit adds the functionality to set GOMAXPROCS in cgroup2 enabled systems.

Also there were lint errors because of build annotation comments which are fixed in a separate commit.

This PR resolves https://github.com/uber-go/automaxprocs/issues/21.

### CODE REVIEW

**Comment by ardan-bkennedy** [comment]
Can someone at Uber provide information about if this project is still being supported? I'm considering to fork the project so it can be maintained.

**Comment by alxn** [comment]
Thanks for the PR - we are looking for a reviewer internally. @prashantv and @akshayjshah have left the building...

**Comment by akshayjshah** [comment]
Happy to look at this PR, Alun, but I'd love to also get a stamp from someone who can speak authoritatively for Uber infra.

**Comment by ardan-bkennedy** [comment]
You don't need to wait.  Run this command at the root of your project where the go.mod file is located

`go get github.com/emadolsky/automaxprocs@master`

Then switch your import to using this.

`"github.com/emadolsky/automaxprocs/maxprocs"`

Everything will work.

**Comment by dprotaso** [comment]
cc @abhinav

**Review by akshayjshah — COMMENTED** [review]
Thanks for the PR, @emadolsky! I'm not at Uber any more and didn't actually author this package, but it's nice to see this package still providing some value for the community.

I've left some comments on the PR (which hopefully the current Uber folks agree with). I'd wait for a review from one of the current maintainers before making any changes, though - in practice, they may not be accepting PRs for this package at all. In the meantime, it looks like you're the maintainer of the new `automaxprocs`! 😸

File: internal/cgroups/cgroups.go
Comment (maciej) [reply internal/cgroups/cgroups.go:187]:
```
  cpu.max
	A read-write two value file which exists on non-root cgroups.
	The default is "max 100000".

	The maximum bandwidth limit.  It's in the following format::

	  $MAX $PERIOD

	which indicates that the group may consume upto $MAX in each
	$PERIOD duration.  "max" for $MAX indicates no limit.  If only
	one number is written, $MAX is updated.
```

https://www.kernel.org/doc/Documentation/admin-guide/cgroup-v2.rst

I think it's safe to assume that if this format were to change the semantics of it would be different, hence this code would need to be revised to support it.

Comment (emadolsky) [reply internal/cgroups/cgroups.go:187]: Thanks for the comments. I've changed the code accordingly.

File: internal/cgroups/cgroups.go
Comment (maciej) [inline internal/cgroups/cgroups.go:175]: Here you should also include the cgroup-path from `/proc/self/cgroup`, see https://man7.org/linux/man-pages/man7/cgroups.7.html

File: internal/cgroups/cgroups2.go
Comment (abhinav) [inline internal/cgroups/cgroups2.go:67]: My knowledge of cgroups is questionable but reading this, I would guess that we shouldn't hard-code the mount point and instead, if the type was "cgroups2", record the mount point and use that path when determining the quota. Is that not desirable?

**Review by abhinav — COMMENTED** [review]
Hey folks, sorry for the delay. @mway and I will review this.
I've made an initial pass in terms of style and organization and made some updates.
In particular, I've moved the bulk of the v2 logic into separate files.

On top of that, I have limited knowledge of the underlying systems so I need some help understanding this:
Why is the mountpoint hard-coded to be /sys/fs/cgroup? As far as I can tell, mountinfo says "this is where cgroups2 is mounted", so should we not use whatever path is specified in mountinfo? Is it absolutely required for cgroups2 to mount to that path?

File: internal/cgroups/cgroups2.go
Comment (emadolsky) [reply internal/cgroups/cgroups2.go:67]: The same thing has been done on runc: https://github.com/opencontainers/runc/blob/main/libcontainer/cgroups/utils.go#L34-L50

If I'm not mistaken, cgroup v2 & v1 can be mounted simultaneously, and in that case, cgroup v2 is mounted on /sys/fs/cgroup/unified. From seeing this code I thought it'd be enough but I can't say for sure.

Comment (abhinav) [reply internal/cgroups/cgroups2.go:67]: Looks like the linked code differentiates between /sys/fs/cgroup and /sys/fs/cgroup/unified.
Looking around a system running in hybrid mode ([per this stack answer](https://unix.stackexchange.com/a/480748)), it seems to me that if we follow my suggestion, we'll need a second check for whether it's in hybrid mode.
I'm okay with hard-coding the path in that case.

**Review by abhinav — APPROVED** [review]
I've made changes for readability, added more tests, and changed cpu_quota_linux so that the changes to the core logic are minimal.

This change LGTM now. I'm going to have someone else take a quick look over it, but it should be good to go otherwise.

**Review by mway — APPROVED** [review]
lgtm

File: internal/cgroups/cgroups2.go
Comment (mway) [inline internal/cgroups/cgroups2.go:116]:
```suggestion

	scanner := bufio.NewScanner(cpuMaxParams)
```

File: internal/cgroups/cgroups2.go
Comment (mway) [inline internal/cgroups/cgroups2.go:150]:
```suggestion
func (cg *CGroups2) CPUQuota() (float64, bool, error) {
    cpuMaxParams, err := os.Open(path.Join(cg.mountPoint, cg.cpuMaxFile))
    if err != nil {
        if os.IsNotExist(err) {
            return -1, false, nil
        }
        return -1, false, err
    }

    scanner := bufio.NewScanner(cpuMaxParams)
    if scanner.Scan() {
        fields := strings.Fields(scanner.Text())
        if len(fields) == 0 || len(fields) > 2 {
            return -1, false, fmt.Errorf("invalid format")
        }

        if fields[_cgroupv2CPUMaxQuotaIndex] == _cgroupV2CPUMaxQuotaMax {
            return -1, false, nil
        }

        max, err := strconv.Atoi(fields[_cgroupv2CPUMaxQuotaIndex])
        if err != nil {
            return -1, false, err
        }

        var period int
        if len(fields) == 1 {
            period = _cgroupV2CPUMaxDefaultPeriod
        } else {
            period, err = strconv.Atoi(fields[_cgroupv2CPUMaxPeriodIndex])
            if err != nil {
                return -1, false, err
            }
        }

        return float64(max) / float64(period), true, nil
    }

    if err := scanner.Err(); err != nil {
        return -1, false, err
    }

    return 0, false, io.ErrUnexpectedEOF
}
```

File: internal/cgroups/cgroups2.go
Comment (mway) [inline internal/cgroups/cgroups2.go:102]:
```suggestion
func isCGroupV2(procPathMountInfo string) (bool, error) {
    var (
        isV2          bool
        newMountPoint = func(mp *MountPoint) error {
            isV2 = mp.FSType == _cgroupv2FSType && mp.MountPoint == _cgroupv2MountPoint
            return nil
        }
    )

    if err := parseMountInfo(procPathMountInfo, newMountPoint); err != nil {
        return false, err
    }

    return isV2, nil
}
```

**Comment by ardan-bkennedy** [comment]
This change has not fixed the problem. I switched my code from "github.com/emadolsky/automaxprocs/maxprocs" to v1.5.0 and I am still getting that the CPU Quota is not defined.

**Comment by ardan-bkennedy** [comment]
I noticed that in v1.5.0
This function is not checking anything related to cgroups2

https://github.com/uber-go/automaxprocs/blob/v1.5.0/internal/runtime/cpu_quota_linux.go#L35
```
func CPUQuotaToGOMAXPROCS(minValue int) (int, CPUQuotaStatus, error) {[Jason Lai, 5 years ago: • Import `automaxprocs/x/runtime`](https://sourcegraph.com/github.com/uber-go/automaxprocs/-/commit/572239b375b1ca6b76babeb8b0ee139dc35fd4d8)
	cgroups, err := newQueryer()
	if err != nil {
		return -1, CPUQuotaUndefined, err
	}

	quota, defined, err := cgroups.CPUQuota()
	if !defined || err != nil {
		return -1, CPUQuotaUndefined, err
	}

	maxProcs := int(math.Floor(quota))
	if minValue > 0 && maxProcs < minValue {
		return minValue, CPUQuotaMinUsed, nil
	}
	return maxProcs, CPUQuotaUsed, nil
}
```

In the emadolsky repo, this function is checking.
```
func CPUQuotaToGOMAXPROCS(minValue int) (int, CPUQuotaStatus, error) {
	var quota float64
	var defined bool
	var err error

	isV2, err := cg.IsCGroupV2()
	if err != nil {
		return -1, CPUQuotaUndefined, err
	}

	if isV2 {
		quota, defined, err = cg.CPUQuotaV2()
		if !defined || err != nil {
			return -1, CPUQuotaUndefined, err
		}
	} else {
		cgroups, err := cg.NewCGroupsForCurrentProcess()
		if err != nil {
			return -1, CPUQuotaUndefined, err
		}

		quota, defined, err = cgroups.CPUQuota()
		if !defined || err != nil {
			return -1, CPUQuotaUndefined, err
		}
	}

	maxProcs := int(math.Floor(quota))
	if minValue > 0 && maxProcs < minValue {
		return minValue, CPUQuotaMinUsed, nil
	}
	return maxProcs, CPUQuotaUsed, nil
}
```

**Comment by bsergean** [comment]
I've add issue getting this working on k8s 1.26 (on google cloud), where apparently we use cgroup v2 ; I'm not positive that this is working.

**Pull request merged**
